substrate/frame/delegated-staking/src/types.rs

```diff
@@ -0,0 +1,308 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Basic types used in delegated staking.
+
+use super::*;
+use frame_support::traits::DefensiveSaturating;
+
+/// The type of pot account being created.
+#[derive(Encode, Decode)]
+pub(crate) enum AccountType {
+	/// A proxy delegator account created for a nominator who migrated to an `Agent` account.
+	///
+	/// Funds for unmigrated `delegator` accounts of the `Agent` are kept here.
+	ProxyDelegator,
+}
+
+/// Information about delegation of a `delegator`.
+#[derive(Default, Encode, Clone, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct Delegation<T: Config> {
+	/// The target of delegation.
+	pub agent: T::AccountId,
+	/// The amount delegated.
+	pub amount: BalanceOf<T>,
+}
+
+impl<T: Config> Delegation<T> {
+	/// Get delegation of a `delegator`.
+	pub(crate) fn get(delegator: &T::AccountId) -> Option<Self> {
+		<Delegators<T>>::get(delegator)
+	}
+
+	/// Create and return a new delegation instance.
+	pub(crate) fn from(agent: &T::AccountId, amount: BalanceOf<T>) -> Self {
+		Delegation { agent: agent.clone(), amount }
+	}
+
+	/// Ensure the delegator is either a new delegator or they are adding more delegation to the
+	/// existing agent.
+	///
+	/// Delegators are prevented from delegating to multiple agents at the same time.
+	pub(crate) fn can_delegate(delegator: &T::AccountId, agent: &T::AccountId) -> bool {
+		Delegation::<T>::get(delegator)
+			.map(|delegation| delegation.agent == agent.clone())
+			.unwrap_or(
+				// all good if it is a new delegator except it should not be an existing agent.
+				!<Agents<T>>::contains_key(delegator),
+			)
+	}
+
+	/// Checked decrease of delegation amount. Consumes self and returns a new copy.
+	pub(crate) fn decrease_delegation(self, amount: BalanceOf<T>) -> Option<Self> {
```

Comment:
admittedly a bit of a strange from a pure Rust persepctive. Consuming self, returning `Option<Self>`, why? 

First comment, the fields of the struct are totally public. Possibly, you can just alter those in place, with checks. 

If you want to go the extra mile, I suggest this somewhat battle-tested FRAME pattern: 

- make the fields private, and make a standard with yourself that you nerver read the storage directly.
- have a `fn get(key)` and `fn put(self)`.
- All other mutable functions would work with `fn foo(&mut self)`

Reply:
I would say the current code is reasonable IFF there is a strong reason where you don't want this item to remain in scope in case the addition fails. That is a good reason to consume `self`, and very Rust-idiomatic. But otherwise it is a bit against all idioms I know of, and at an optimization nitpick it has needless allocation to move objects around instead of refs.

Reply:
Yeah not sure I was trying to do but removed these and added inline sub.